java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -85,55 +142,78 @@ void buildMapping(TypeDescription fileType,
     // check the easy case first
     if (fileType.getCategory() == readerType.getCategory()) {
       switch (readerType.getCategory()) {
-        case BOOLEAN:
-        case BYTE:
-        case SHORT:
-        case INT:
-        case LONG:
-        case DOUBLE:
-        case FLOAT:
-        case STRING:
-        case TIMESTAMP:
-        case BINARY:
-        case DATE:
-          // these are always a match
-          break;
-        case CHAR:
-        case VARCHAR:
-          // HIVE-13648: Look at ORC data type conversion edge cases (CHAR, VARCHAR, DECIMAL)
-          isOk = fileType.getMaxLength() == readerType.getMaxLength();
-          break;
-        case DECIMAL:
-          // HIVE-13648: Look at ORC data type conversion edge cases (CHAR, VARCHAR, DECIMAL)
-          // TODO we don't enforce scale and precision checks, but probably should
-          break;
-        case UNION:
-        case MAP:
-        case LIST: {
-          // these must be an exact match
-          List<TypeDescription> fileChildren = fileType.getChildren();
-          List<TypeDescription> readerChildren = readerType.getChildren();
-          if (fileChildren.size() == readerChildren.size()) {
-            for(int i=0; i < fileChildren.size(); ++i) {
-              buildMapping(fileChildren.get(i), readerChildren.get(i));
-            }
-          } else {
-            isOk = false;
+      case BOOLEAN:
+      case BYTE:
+      case SHORT:
+      case INT:
+      case LONG:
+      case DOUBLE:
+      case FLOAT:
+      case STRING:
+      case TIMESTAMP:
+      case BINARY:
+      case DATE:
+        // these are always a match
+        break;
+      case CHAR:
+      case VARCHAR:
+        // HIVE-13648: Look at ORC data type conversion edge cases (CHAR, VARCHAR, DECIMAL)
+        isOk = fileType.getMaxLength() == readerType.getMaxLength();
+        break;
+      case DECIMAL:
+        // HIVE-13648: Look at ORC data type conversion edge cases (CHAR, VARCHAR, DECIMAL)
+        // TODO we don't enforce scale and precision checks, but probably should
+        break;
+      case UNION:
+      case MAP:
+      case LIST: {
+        // these must be an exact match
+        List<TypeDescription> fileChildren = fileType.getChildren();
+        List<TypeDescription> readerChildren = readerType.getChildren();
+        if (fileChildren.size() == readerChildren.size()) {
+          for (int i = 0; i < fileChildren.size(); ++i) {
+            buildMapping(fileChildren.get(i), readerChildren.get(i), useFieldNames);
```

Comment:
when you are recursing, useFieldNames should always be true. Prior to HIVE-4243 only the top level column names were lost. In recursive types the field names were correct.